### PR TITLE
Update pinnings

### PR DESCRIPTION
**Pull request by ocefpaf**
~~@pelson I am not really sure if these are needed.~~

### CODE REVIEW

**Comment by ocefpaf** [comment]
I used this PR to add `netcdf-cxx4` and `netcdf-fortran`, and update the pinnings to match what we have as latest. Merge with caution as some discussion is needed before we enforce this pinnings.

**Comment by jjhelmus** [comment]
I'm in favor of these, although I am only worried about updates to the `xz` and `sqlite` pins.

A few notes:
* `readline` **6.3** is out and available in the `conda-forge` channel but `defaults` is still pinning **6.2**
* `tk` **8.6.5** is out but the latest version in the `conda-forge` channel is **8.5.19**.  The `defaults` channel seems to be pinning **8.5.18** which is the latest version available from that channel and is in agreement with the current and proposed pinnings.
* `ncurses` **6.0** is out but both the `conda-forge` and `defaults` channels provide **5.9**.

**Comment by jakirkham** [comment]
We were trying to get `tk` 8.6.5 done, but it is having issues ( https://github.com/conda-forge/tk-feedstock/pull/1 ).

I think we should investigate the implications of `ncurses` 6.0 (degree of breakage). There are also some packages that are not pinned correctly floating around in conda-forge. So, I held off. ( https://github.com/conda-forge/ncurses-feedstock/issues/14 ) The feedstocks should all be pinned correctly now or at least have been provided the opportunity.

File: scripts/pin_the_slow_way.py [hunk 1]
```diff
@@ -37,14 +37,16 @@
           'icu': 'icu 56.*',
           'jpeg': 'jpeg 9*',
           'libnetcdf': 'libnetcdf 4.4.*',
-          'libpng': 'libpng >=1.6.21,<1.7',
+          'libpng': 'libpng 1.6.23',
```
Comment (pelson) [inline scripts/pin_the_slow_way.py]: Was there an API change which broke libpng in 23 as well? :cry:

**Comment by pelson** [comment]
I'm confident in saying that this will tease out futher issues with the conda-forge stack (which other pinned version changes haven't! :wink:). Though I'm in two minds as to whether I want to aim for some stability for the next few weeks (SciPy is on, and I'll be there -> lots of new users, and me not being fully available to fix any problems).

File: scripts/pin_the_slow_way.py (hunk 1 above)
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: Not sure to be honest. But I got some reports of breaks in libpng `1.6.21` (conda-forge) and `1.6.22` (`defaults`)

**Comment by pelson** [comment]
In principle, I'm very happy with moving these forwards. In practice, I could see an argument for doing this now, but can also see an argument for getting some stability. Are there drivers for doing this to solve existing issues?

**Comment by ocefpaf** [comment]
>SciPy is on, and I'll be there -> lots of new users, and me not being fully available to fix any problems

By no means I want this merge soon nor `MTN` PRs updating this. Maybe we should only do that once we figure out a better way to release new packages in an unstable channel first.

> Are there drivers for doing this to solve existing issues?

Mostly is to update and move forward. However, `xz` **is** now a real problem as `defaults` moved to `5.2.*` and `xz` is tied to `Python 3.x`. Meaning that the solver is updating Python and `xz` from `defaults` and breaking/not-resolving everything that is pinned to `xz 5.0.*`.

**Comment by pelson** [comment]
Thanks @ocefpaf. We can certainly separate updating `xz`. That does sound like an important step to keep up to speed with the latest anaconda.

**Comment by ocefpaf** [comment]
> Thanks @ocefpaf. We can certainly separate updating `xz`

See https://github.com/conda-forge/staged-recipes/issues/907#issuecomment-229152492 for one example where that is important. (I can prepare another PR here just `xz` later today if you don't beat me to it.)

**Comment by jjhelmus** [comment]
The defaults channel's python-3.x also pins a more recent `sqlite` than our current pinnings (`3.13.*` vs `3.9.*`)

From the os-x python-3.5.1-5 package:

```
        - bzip2
        - zlib 1.2.*
        - sqlite 3.13.*
        - readline 6.2
        - tk 8.5.18
        - openssl 1.0.2*
        - xz 5.2.*
```

I would be happy with a PR that just updated `xz` and `sqlite`.  A `conda-forge` Python 3.5 recipe is already read with these versions, conda-forge/python-feedstock#26

**Comment by ocefpaf** [comment]
> The defaults channel's python-3.x also pins a more recent `sqlite` than our current pinnings (`3.13.*` vs `3.9.*`)

Agreed. See https://github.com/conda-forge/conda-forge.github.io/pull/182

**Comment by jakirkham** [comment]
> * `readline` **6.3** is out and available in the `conda-forge` channel but `defaults` is still pinning **6.2**

Should have added we also pin to 6.2.

**Comment by jjhelmus** [comment]
> Should have added we also pin to 6.2.

`defaults` seems to be pinning to **6.2** also so this is not an issue.  Although trying to upgrade readline in a conda environment can have some interesting results (although conda upgrade --all works fine):

```
$ conda upgrade -c conda-forge readline
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment /home/jhelmus/anaconda/envs/readline_test:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python-1.2                 |                0         773 KB  conda-forge
    readline-6.3               |                1         787 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         1.5 MB

The following packages will be UPDATED:

    readline: 6.2-0   conda-forge --> 6.3-1 conda-forge

The following packages will be DOWNGRADED due to dependency conflicts:

    python:   3.5.2-1 conda-forge --> 1.2-0 conda-forge

Proceed ([y]/n)? 
```

**Comment by jakirkham** [comment]
> ```
> The following packages will be DOWNGRADED due to dependency conflicts:
>
>    python:   3.5.2-1 conda-forge --> 1.2-0 conda-forge
> ```

😱

File: scripts/pin_the_slow_way.py (hunk 1 above)
Comment (jakirkham) [inline scripts/pin_the_slow_way.py]: Could you please provide references?

**Comment by jakirkham** [comment]
Has some merge conflicts that need to be resolved.

File: scripts/pin_the_slow_way.py (hunk 1 above)
Comment (183amir) [inline scripts/pin_the_slow_way.py]: Please see this issue: https://github.com/conda-forge/matplotlib-feedstock/issues/41

Comment (183amir) [inline scripts/pin_the_slow_way.py]: Please see here too: https://github.com/conda-forge/libpng-feedstock/issues/10

**Comment by pelson** [comment]
@ocefpaf - would you mind rebasing this - I think we can go ahead and merge.

**Comment by ocefpaf** [comment]
I am on the road but I'll try to connect later today and rebase. (It is
almost a miracle to get 3G coverage from where I am today :-)

File: scripts/pin_the_slow_way.py [hunk 2]
```diff
@@ -39,10 +39,12 @@
           'libblitz': 'libblitz 0.10|0.10.*',
           'libmatio': 'libmatio 1.5.*',
           'libnetcdf': 'libnetcdf 4.4.*',
-          'libpng': 'libpng >=1.6.21,<1.7',
+          'libpng': 'libpng >=1.6.23,<1.7',
```
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: See https://github.com/conda-forge/matplotlib-feedstock/pull/52#issuecomment-240448503

**Comment by ocefpaf** [comment]
> @ocefpaf - would you mind rebasing this - I think we can go ahead and merge.

Done. I will be online for a couple of hours. Let me know if you need anything else.

**Comment by 183amir** [comment]
Hey guys, please please pin libpng on a patch number. I beg of you. I have been having trouble with this for so long but you don't listen. Please pin it on 1.6.22 too to be inline with the defaults channel.

**Comment by ocefpaf** [comment]
> Hey guys, please please pin libpng on a patch number. I beg of you. I have been having trouble with this for so long but you don't listen. Please pin it on 1.6.22 too to be inline with the defaults channel.

There is no need for desperation. See https://github.com/conda-forge/conda-forge.github.io/pull/211#issuecomment-250740925

The context of this PR here is not to fix that. I am just waiting for @pelson's comment on https://github.com/conda-forge/conda-forge.github.io/pull/211#issuecomment-250740925 to rebase and pin there.

**Comment by 183amir** [comment]
Here is a log that shows matplotlib compiled with 1.6.24 does not work with libpng 1.6.23 https://travis-ci.org/conda-forge/bob.ip.facedetect-feedstock/jobs/163096191

File: scripts/pin_the_slow_way.py (hunk 2 above)
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: This is more restrictive (an working in some cases) than the current setup. So I am merging this to get the other changes in too.  We have another PR that pins `libpng` to the bone and any further discussion should go there.

See https://github.com/conda-forge/conda-forge.github.io/pull/211

**Pull request merged**